src/storage/src/sink/kafka.rs

```diff
@@ -336,16 +365,40 @@ impl TransactionalProducer {
             timestamp: None,
             delivery_opaque: (),
         };
-        let key_size = key.map(|k| k.len()).unwrap_or(0);
-        let value_size = value.map(|k| k.len()).unwrap_or(0);
-        let record_size = u64::cast_from(key_size + value_size);
-        self.statistics.inc_messages_staged_by(1);
-        self.staged_messages += 1;
-        self.statistics.inc_bytes_staged_by(record_size);
-        self.staged_bytes += record_size;
+
         self.producer.send(record).map_err(|(e, _)| e)
     }
 
+    /// Updates the reported statistics for this sink based on the statistics received from
+    /// librdkafka. Since `librdkafka` does not differentiate between committed and uncomitted
+    /// updates the function accepts a flag to indicate if it should assume that the messages
+    /// reported as produced are also committed.
+    fn update_stats(&mut self, assume_committed: bool) {
+        let stats = self.stats_receiver.borrow_and_update();
+        let messages_staged = u64::try_from(stats.txmsgs).unwrap();
+        let bytes_staged = u64::try_from(stats.txmsg_bytes).unwrap();
+
+        self.statistics.set_messages_buffered(stats.msg_cnt);
+        self.statistics
+            .inc_messages_staged_by(messages_staged - self.prev_messages_staged);
+        self.prev_messages_staged = messages_staged;
+
+        self.statistics.set_bytes_buffered(stats.msg_size);
+        self.statistics
+            .inc_bytes_staged_by(bytes_staged - self.prev_bytes_staged);
+        self.prev_bytes_staged = bytes_staged;
+
+        if assume_committed {
```

Comment:
On a first pass I don't see how this works out. Assume I call this method twice, first with `assume_committed=false` and later `=true`. It seems to me that we'd increment the "committed" stats by 0? And since you have the stats-reporting thread racing with the actual operation, it seems really hard to predict what these numbers would look like.

As a kneejerk I'd be tempted to just drop these derived metrics, report on the raw data from `rdkafka`, and do any transformations we need in Grafana. But I don't know well enough how these metrics are used to be confident that this makes sense...

Reply:
>Assume I call this method twice, first with assume_committed=false and later =true. It seems to me that we'd increment the "committed" stats by 0?

that's right, we call it with `true` immediately after `commit_transaction` has finished and before any more messages are sent. It doesn't end up incrementing the committed stats by zero because we track the previous value reported for both the staged and the committed variants. 

> And since you have the stats-reporting thread racing with the actual operation, it seems really hard to predict what these numbers would look like.

Yeah.. I reached the same realization but I couldn't find a way to both have committed stats and not have this annoying race. The thing that gave me peace is that we commit transactions often and so even if we miss the race we'll get the updated stats a second later, since the stats are being updated every 500ms.

> As a kneejerk I'd be tempted to just drop these derived metrics, report on the raw data from rdkafka, and do any transformations we need in Grafana.

The only alternative that comes to mind is to somehow record the timestamp of the last committed transaction as a metric and then have Grafana calculate the sum of staged  metric until that timestamp, if that's even possible. Did you have something else in mind?